Feature/465 directed graphsage

geoffj-d61 opened the pull request:
Updated the GraphSAGE aggregation architecture to create a directed GraphSAGE algorithm that separately handles in-node and out-node neighbourhoods. Some (temporary) demo examples have been added to check the differences in behaviour between directed and undirected GraphSAGE. 

Note that example 3 (directed GraphSAGE, no in-nodes, only out-nodes) is now broken; this is not because of directionality, but because of an existing bug in the framework: the mean of a tensor with a 0 dimension produces NaNs. Undirected GraphSAGE suffers this same bug (try setting the 2nd sampling dimension to 0 in either example 1 or 2). Will need help to fix this.

adocherty commented:
OK, my changes have broken the unit tests. This is due to the `AttentionalAggregator` which doesn't fit into the paradigm of the other aggregators. I'll commit a fix tomorrow 😟

geoffj-d61 commented:
@adocherty I have moved example4 into graphsage, as you suggested, and deleted directed-graphsage. The example and experimental notebooks have been saved as branch feature/465-directed-graphsage on stellar-research.

geoffj-d61 commented on stellargraph/layer/graphsage.py:
```diff
@@ -127,8 +135,39 @@ def build(self, input_shape):
+        for neigh_shape in input_shape[1:]:
+            if neigh_shape[2] == 0:
+                counter = counter + 1
+        self._build_mlp_only = counter == len(input_shape)
```
Method _build_mlp_only() removed.

geoffj-d61 commented on stellargraph/layer/graphsage.py:
```diff
@@ -302,7 +352,7 @@ def build(self, input_shape):
             )
 
-    def aggregate_neighbours(self, x_neigh):
+    def aggregate_neighbours(self, x_neigh, **kwargs):
```
Fixed by @adocherty .

geoffj-d61 commented on stellargraph/layer/graphsage.py:
```diff
@@ -747,3 +806,198 @@ def default_model(self, flatten_output=True):
+
+    def _build_aggregators(self):
+        # Model parameters
+        self.max_hops = max_hops = self.n_layers
```
Renamed n_layers to max_slots.

geoffj-d61 commented on stellargraph/layer/graphsage.py:
```diff
@@ -747,3 +806,198 @@ def default_model(self, flatten_output=True):
+        out_layer = Reshape(K.int_shape(out_layer)[2:])(out_layer)
+        return self._normalization(out_layer)
+
+    def _compute_input_sizes(self) -> List[int]:
```
Added call from `GraphSAGE.__init__()` to `_compute_neighbourhood_sizes()` to set `self.neighbourhood_sizes` (and `max_slots` for `DirectedGraphSAGE`). Updated `GraphSAGE.node_model()` to use `self.neighbourhood_sizes`, and removed `DirectedGraphSAGE.node_model()`.

geoffj-d61 commented on stellargraph/layer/graphsage.py:
```diff
@@ -747,3 +806,198 @@ def default_model(self, flatten_output=True):
+                "If generator is not provided, in_samples, out_samples and input_dim must be specified."
+            )
+
+    def _build_aggregators(self):
```
- Shifted setting of `max_slots` to `DirectedGraphSAGE._compute_neighbourhood_sizes()`. 

- Shifted checking of `in_samples` and `out_samples` lengths to `_get_sizes_from_generator()` and `_get_sizes_from_keywords()`. 

- Added similar checking to `GraphSAGE` versions. 

- Removed redundant `DirectedGraphSAGE._build_aggregators()` method.

geoffj-d61 commented on stellargraph/data/explorer.py:
```diff
@@ -1008,106 +708,144 @@ def run(self, nodes=None, n=1, n_size=None, seed=None):
 
-    def _check_parameter_values(self, nodes, n, n_size, graph_schema, seed):
+
+class DirectedBreadthFirstNeighbours(GraphWalk):
```
I guess this refactoring will be another issue.

geoffj-d61 commented on stellargraph/layer/graphsage.py:
```diff
@@ -747,3 +806,198 @@ def default_model(self, flatten_output=True):
+        out_layer = stage_tree[0]
+
+        # Remove neighbourhood dimension from output tensors of the stack
+        out_layer = Reshape(K.int_shape(out_layer)[2:])(out_layer)
```
@adocherty what is `x` here in `if K.int_shape(x)==1`? The method input is `xin` which is the (flattened binary tree) list of tensors, so it's not that. And the `out_layer` we are expecting to have 2 or more dimensions. So I'm a bit confused here!

geoffj-d61 replied:
@adocherty I think the above is the last outstanding issue.

adocherty replied:
Hi, yes I meant `K.int_shape(out_layer)[1] == 1`

geoffj-d61 replied:
Added and committed!

geoffj-d61 commented:
The one thing I forgot to check fully was the demo, which got broken by the keras -> tensorflow.keras changes! Fixed now.
